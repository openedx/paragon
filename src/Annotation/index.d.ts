import React from 'react';

export interface AnnotationProps {
  className?: string,
  variant?: 'error' | 'success' | 'warning' | 'light' | 'dark',
  children?: React.ReactNode,
  arrowPlacement?: 'top' | 'right' | 'bottom' | 'left',
}
declare const Annotation: React.ForwardRefExoticComponent<AnnotationProps>;
export default Annotation;
