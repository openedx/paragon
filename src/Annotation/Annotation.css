/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:19 GMT
 */

:root {
  --pgn-spacing-annotation-arrow-side-margin: .25rem;
  --pgn-spacing-annotation-padding: .5rem;
  --pgn-size-annotation-border-radius: .25rem;
  --pgn-size-annotation-max-width: 18.75rem;
  --pgn-size-annotation-arrow-border-width: .5rem;
  --pgn-typography-annotation-line-height: var(--pgn-typography-line-height-sm);
  --pgn-typography-annotation-font-size: var(--pgn-typography-font-size-sm);
}
