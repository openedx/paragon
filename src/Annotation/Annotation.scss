@import "variables";

@mixin triangle($color, $direction) {
  height: 0;
  width: 0;
  position: absolute;
  content: "";
  border: solid transparent;

  @if $direction == top {
    border-bottom-color: $color;
    border-width: 0 $annotation-arrow-border-width $annotation-arrow-border-width;
    right: 0;
    left: 0;
    top: -$annotation-arrow-border-width;
    margin: 0 auto;
  }

  @else if $direction == right {
    border-left-color: $color;
    border-width: $annotation-arrow-border-width 0 $annotation-arrow-border-width $annotation-arrow-border-width;
    top: 0;
    bottom: 0;
    right: -$annotation-arrow-border-width;
    margin: auto 0;
  }

  @else if $direction == bottom {
    border-top-color: $color;
    border-width: $annotation-arrow-border-width $annotation-arrow-border-width 0;
    right: 0;
    left: 0;
    bottom: -$annotation-arrow-border-width;
    margin: 0 auto;
  }

  @else if $direction == left {
    border-right-color: $color;
    border-width: $annotation-arrow-border-width $annotation-arrow-border-width $annotation-arrow-border-width 0;
    top: 0;
    bottom: 0;
    left: -$annotation-arrow-border-width;
    margin: auto 0;
  }

  @else {
    @error "Unknown direction #{$direction}.";
  }
}

.pgn__annotation {
  padding: $annotation-padding;
  border-radius: $annotation-border-radius;
  max-width: $annotation-max-width;
  filter: $annotation-box-shadow;
  word-wrap: break-word;
  position: relative;
  display: inline-block;
  font-size: $annotation-font-size;
  line-height: $annotation-line-height;
  text-align: center;
}

$arrow-directions: top, right, bottom, left;

@each $name, $colors in $annotation-variants {
  @each $direction in $arrow-directions {
    .pgn__annotation-#{$name}-#{$direction} {
      background-color: map-get($colors, "background-color");
      color: map-get($colors, "color");

      // set additional margin to arrow side of the Annotation
      margin-#{$direction}: $annotation-arrow-border-width + $annotation-arrow-side-margin;

      [dir="rtl"] & {
        @if $direction == left {
          margin-left: 0;
          margin-right: $annotation-arrow-border-width + $annotation-arrow-side-margin;
        }

        @else if $direction == right {
          margin-right: 0;
          margin-left: $annotation-arrow-border-width + $annotation-arrow-side-margin;
        }
      }

      &::after {
        @include triangle(map-get($colors, "background-color"), $direction);

        [dir="rtl"] & {
          @if $direction == left {
            left: initial;
            right: -$annotation-arrow-border-width;
            border-width:
              $annotation-arrow-border-width 0 $annotation-arrow-border-width
              $annotation-arrow-border-width;
            border-left-color: map-get($colors, "background-color");
          }

          @else if $direction == right {
            right: initial;
            left: -$annotation-arrow-border-width;
            border-width:
              $annotation-arrow-border-width $annotation-arrow-border-width
              $annotation-arrow-border-width 0;
            border-right-color: map-get($colors, "background-color");
          }
        }
      }
    }
  }
}
