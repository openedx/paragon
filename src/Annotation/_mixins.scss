@mixin annotation-triangle($triangle-color, $triangle-direction) {
  content: "";
  height: 0;
  width: 0;
  position: absolute;
  border: solid transparent;

  @if $triangle-direction == top {
    border-bottom-color: $triangle-color;
    border-width: 0 var(--pgn-size-annotation-arrow-border-width) var(--pgn-size-annotation-arrow-border-width);
    right: 0;
    left: 0;
    top: calc(var(--pgn-size-annotation-arrow-border-width) * -1);
    margin: 0 auto;
  }

  @else if $triangle-direction == right {
    border-left-color: $triangle-color;
    border-width:
      var(--pgn-size-annotation-arrow-border-width) 0 var(--pgn-size-annotation-arrow-border-width)
      var(--pgn-size-annotation-arrow-border-width);
    top: 0;
    bottom: 0;
    right: calc(var(--pgn-size-annotation-arrow-border-width) * -1);
    margin: auto 0;
  }

  @else if $triangle-direction == bottom {
    border-top-color: $triangle-color;
    border-width: var(--pgn-size-annotation-arrow-border-width) var(--pgn-size-annotation-arrow-border-width) 0;
    right: 0;
    left: 0;
    bottom: calc(var(--pgn-size-annotation-arrow-border-width) * -1);
    margin: 0 auto;
  }

  @else if $triangle-direction == left {
    border-right-color: $triangle-color;
    border-width:
      var(--pgn-size-annotation-arrow-border-width) var(--pgn-size-annotation-arrow-border-width)
      var(--pgn-size-annotation-arrow-border-width) 0;
    top: 0;
    bottom: 0;
    left: calc(var(--pgn-size-annotation-arrow-border-width) * -1);
    margin: auto 0;
  }

  @else {
    @error "Unknown direction #{$triangle-direction}.";
  }
}

@mixin annotation-variant($bg-color, $text-color, $direction) {
  background-color: $bg-color;
  color: $text-color;

  // set additional margin to arrow side of the Annotation
  margin-#{$direction}:
    calc(
      var(--pgn-size-annotation-arrow-border-width) + var(--pgn-spacing-annotation-arrow-side-margin)
    );

  [dir="rtl"] & {
    @if $direction == left {
      margin-left: 0;
      margin-right:
        calc(
          var(--pgn-size-annotation-arrow-border-width) + var(--pgn-spacing-annotation-arrow-side-margin)
        );
    }

    @else if $direction == right {
      margin-right: 0;
      margin-left:
        calc(
          var(--pgn-size-annotation-arrow-border-width) + var(--pgn-spacing-annotation-arrow-side-margin)
        );
    }
  }

  &::after {
    @include annotation-triangle($bg-color, $direction);

    [dir="rtl"] & {
      @if $direction == left {
        left: initial;
        right: calc(var(--pgn-size-annotation-arrow-border-width) * -1);
        border-width:
          var(--pgn-size-annotation-arrow-border-width) 0 var(--pgn-size-annotation-arrow-border-width)
          var(--pgn-size-annotation-arrow-border-width);
        border-left-color: $bg-color;
      }

      @else if $direction == right {
        right: initial;
        left: calc(var(--pgn-size-annotation-arrow-border-width) * -1);
        border-width:
          var(--pgn-size-annotation-arrow-border-width) var(--pgn-size-annotation-arrow-border-width)
          var(--pgn-size-annotation-arrow-border-width) 0;
        border-right-color: $bg-color;
      }
    }
  }
}
