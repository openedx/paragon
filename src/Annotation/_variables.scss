$annotation-padding:               .5rem !default;
$annotation-box-shadow:            drop-shadow(0 2px 4px rgba(0, 0, 0, .15))
  drop-shadow(0 2px 8px rgba(0, 0, 0, .15)) !default;
$annotation-border-radius:         .25rem !default;
$annotation-max-width:             18.75rem !default;
$annotation-font-size:             $font-size-sm !default;
$annotation-line-height:           $line-height-sm !default;

$annotation-variants: (
  "success": ( "background-color": $success, "color": $white ),
  "warning": ( "background-color": $accent-b, "color": $black ),
  "error":   ( "background-color": $danger, "color": $white ),
  "light":   ( "background-color": $white, "color": $primary-500 ),
  "dark":    ( "background-color": $dark, "color": $white ),
) !default;

$annotation-arrow-side-margin:     .25rem !default;
$annotation-arrow-border-width:    .5rem !default;
