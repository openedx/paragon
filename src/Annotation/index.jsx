import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const Annotation = React.forwardRef(({
  className,
  variant,
  children,
  arrowPlacement,
}, ref) => (
  <span
    className={classNames(
      className,
      'pgn__annotation',
      `pgn__annotation-${variant}-${arrowPlacement}`,
    )}
    ref={ref}
  >
    {children}
  </span>
));

Annotation.defaultProps = {
  className: undefined,
  variant: 'success',
  arrowPlacement: 'bottom',
};

Annotation.propTypes = {
  /** Specifies contents of the component. */
  children: PropTypes.node.isRequired,
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** Specifies variant to use. */
  variant: PropTypes.oneOf(['error', 'success', 'warning', 'light', 'dark']),
  /** Specifies arrow position. */
  arrowPlacement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
};

export default Annotation;
