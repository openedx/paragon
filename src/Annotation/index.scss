@import "mixins";

:root {
  --pgn-elevation-annotation-box-shadow:
    drop-shadow(
      var(--pgn-elevation-annotation-box-shadow-1-offset-x)
      var(--pgn-elevation-annotation-box-shadow-1-offset-y)
      var(--pgn-elevation-annotation-box-shadow-1-blur)
      var(--pgn-elevation-annotation-box-shadow-1-color)
    )
    drop-shadow(
      var(--pgn-elevation-annotation-box-shadow-2-offset-x)
      var(--pgn-elevation-annotation-box-shadow-2-offset-y)
      var(--pgn-elevation-annotation-box-shadow-2-blur)
      var(--pgn-elevation-annotation-box-shadow-2-color)
    );
}

.pgn__annotation {
  padding: var(--pgn-spacing-annotation-padding);
  border-radius: var(--pgn-size-annotation-border-radius);
  max-width: var(--pgn-size-annotation-max-width);
  filter: var(--pgn-elevation-annotation-box-shadow);
  word-wrap: break-word;
  position: relative;
  display: inline-block;
  font-size: var(--pgn-typography-annotation-font-size);
  line-height: var(--pgn-typography-annotation-line-height);
  text-align: center;
}

// SUCCESS
.pgn__annotation-success-top {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-success),
    var(--pgn-color-annotation-text-success), "top");
}

.pgn__annotation-success-bottom {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-success),
    var(--pgn-color-annotation-text-success), "bottom");
}

.pgn__annotation-success-right {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-success),
    var(--pgn-color-annotation-text-success), "right");
}

.pgn__annotation-success-left {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-success),
    var(--pgn-color-annotation-text-success), "left");
}

// WARNING
.pgn__annotation-warning-top {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-warning),
    var(--pgn-color-annotation-text-warning), "top");
}

.pgn__annotation-warning-bottom {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-warning),
    var(--pgn-color-annotation-text-warning), "bottom");
}

.pgn__annotation-warning-right {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-warning),
    var(--pgn-color-annotation-text-warning), "right");
}

.pgn__annotation-warning-left {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-warning),
    var(--pgn-color-annotation-text-warning), "left");
}

// ERROR
.pgn__annotation-error-top {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-error),
    var(--pgn-color-annotation-text-error), "top");
}

.pgn__annotation-error-bottom {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-error),
    var(--pgn-color-annotation-text-error), "bottom");
}

.pgn__annotation-error-right {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-error),
    var(--pgn-color-annotation-text-error), "right");
}

.pgn__annotation-error-left {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-error),
    var(--pgn-color-annotation-text-error), "left");
}

// DARK
.pgn__annotation-dark-top {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-dark),
    var(--pgn-color-annotation-text-dark), "top");
}

.pgn__annotation-dark-bottom {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-dark),
    var(--pgn-color-annotation-text-dark), "bottom");
}

.pgn__annotation-dark-right {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-dark),
    var(--pgn-color-annotation-text-dark), "right");
}

.pgn__annotation-dark-left {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-dark),
    var(--pgn-color-annotation-text-dark), "left");
}

// LIGHT
.pgn__annotation-light-top {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-light),
    var(--pgn-color-annotation-text-light), "top");
}

.pgn__annotation-light-bottom {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-light),
    var(--pgn-color-annotation-text-light), "bottom");
}

.pgn__annotation-light-right {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-light),
    var(--pgn-color-annotation-text-light), "right");
}

.pgn__annotation-light-left {
  @include annotation-variant(
    var(--pgn-color-annotation-bg-light),
    var(--pgn-color-annotation-text-light), "left");
}
