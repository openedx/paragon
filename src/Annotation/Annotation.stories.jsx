import Annotation from './index';

export default {
  title: 'Annotation',
  component: Annotation,
  argTypes: {
    arrowPlacement: {
      control: {
        type: 'select',
      },
    },
  },
};

const Template = (args) => <Annotation {...args} />;

export const Error = Template.bind({});
Error.args = { children: 'Error', variant: 'error' };

export const Success = Template.bind({});
Success.args = { children: 'Success', variant: 'success' };

export const Warning = Template.bind({});
Warning.args = { children: 'Warning', variant: 'warning' };

export const Light = Template.bind({});
Light.args = { children: 'Light', variant: 'light' };

export const Dark = Template.bind({});
Dark.args = { children: 'Dark', variant: 'dark' };
