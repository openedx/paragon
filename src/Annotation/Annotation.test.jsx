import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Annotation from './index';

const VARIANTS = ['error', 'success', 'warning', 'light', 'dark'];
const ARROW_PLACEMENTS = ['top', 'right', 'bottom', 'left'];

describe('Annotation', () => {
  it('renders in all variants', () => {
    const tree = renderer.create((
      <>
        {VARIANTS.map((variant) => ARROW_PLACEMENTS.map(position => (
          <Annotation
            key={`${variant}-${position}`}
            arrowPlacement={position}
            variant={variant}
          >
            Annotation text
          </Annotation>
        )))}
      </>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  const classNameTestData = [];
  VARIANTS.forEach((variant) => {
    ARROW_PLACEMENTS.forEach((arrowPlacement) => {
      classNameTestData.push([variant, arrowPlacement, `pgn__annotation-${variant}-${arrowPlacement}`]);
    });
  });

  test.each(classNameTestData)(
    'renders with correct className for variant: %s, with arrow placement: %s',
    (variant, arrowPlacement, expectedClassName) => {
      const wrapper = mount(<Annotation variant={variant} arrowPlacement={arrowPlacement}>Test text</Annotation>);
      const annotation = wrapper.find('.pgn__annotation');
      expect(annotation.hasClass(expectedClassName)).toEqual(true);
    },
  );
});
