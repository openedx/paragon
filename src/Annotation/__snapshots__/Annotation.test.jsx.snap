// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Annotation renders in all variants 1`] = `
Array [
  <span
    className="pgn__annotation pgn__annotation-error-top"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-error-right"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-error-bottom"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-error-left"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-success-top"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-success-right"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-success-bottom"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-success-left"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-warning-top"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-warning-right"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-warning-bottom"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-warning-left"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-light-top"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-light-right"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-light-bottom"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-light-left"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-dark-top"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-dark-right"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-dark-bottom"
  >
    Annotation text
  </span>,
  <span
    className="pgn__annotation pgn__annotation-dark-left"
  >
    Annotation text
  </span>,
]
`;
